### ISSUE TITLE
False positive on `-Wzero-as-null-pointer-constant` with spaceship operator

### ISSUE THREAD

**Issue by brevzin** [llvm/llvm-project#55919]
Reduced from [StackOverflow](https://stackoverflow.com/q/72533435/2069064):

```cpp
#include <compare>

struct quantity {
    int i;
    constexpr auto operator<=>( const quantity& ) const = default;
};

bool f(quantity a, quantity b) {
    return a < b;
}
```

This code, when compiled with `-Wzero-as-null-pointer-constant`, [emits](https://godbolt.org/z/3jvKG67fh):

```
<source>:5:65: error: zero as null pointer constant [-Werror,-Wzero-as-null-pointer-constant]
    constexpr auto operator<=>( const quantity& ) const = default;
                                                                ^
                                                                nullptr
<source>:9:14: note: in defaulted three-way comparison operator for 'quantity' first required here
    return a < b;
             ^
```

But the fact that `a < b` evaluates as `(a <=> b) < 0` and the implementation uses a pointer type for the `0` part of this expression is an implementation detail that the user can't control, this isn't something that should be warned on. It's the language that puts in the `0` and the language that more or less requires that it be implemented as a pointer.

**Comment by jwakely** [comment]
FWIW GCC used to warn the same way https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95242

**Comment by davidstone** [comment]
Duplicate of https://github.com/llvm/llvm-project/issues/43670
